### DISCUSSION TITLE
Why remove support for Tcl/Tk 8.4 when CPython 3.10 supports that version?

### DISCUSSION THREAD

**Discussion by RasPython3** [python-pillow/Pillow#8477] [Q&A] [answered]
I leaned that Pillow removed support for Tcl/Tk 8.4 at version 10.0.0.

However, CPython 3.10 says in [Modules/_tkinter.c](https://github.com/python/cpython/blob/3.10/Modules/_tkinter.c):
> Only Tcl/Tk 8.4 and later are supported. Older versions are not
> supported. Use Python 3.4 or older if you cannot upgrade your
> Tcl/Tk libraries.

This means that at some environment, Pillow 11 won't work well with Tkinter module on Python 3.10.

In my case, I was running CPython 3.10 on Windows CE. When I tried to show PNG image on Tkinter canvas, no picture appered.
At first, I had no idea why it didn't work, because there was no errors.
When I checked Tcl/Tk document, I noticed that Tk_PhotoPutBlock takes additional arguments after Tk 8.5, and that latest Pillow doesn't support Tcl/Tk 8.4.
Tcl/Tk on my device was version 8.4, which is the latest version of Tcl/Tk ported to WinCE.

I agree that Tcl/Tk 8.4 is already out of fashion, but isn't it proper to support Tcl/Tk 8.4 as long as Pillow supports Python 3.10?

(My English might be wrong because I'm not native English speaker, sorry.)

**Comment by radarhere** [comment, marked as answer]
This might be obvious, but it states that 'Tcl/Tk 8.4 **and later** are supported'. So Python 3.10 should work fine with Tcl/Tk 8.5.

[My thinking](https://github.com/python-pillow/Pillow/pull/5216#issue-791016259) was that Tcl 8.5 was released in 2007. So everything has had 15 years to upgrade from Tcl 8.4 before Pillow removed support for it.

The Tcl website, https://core.tcl-lang.org/tcl/wiki?name=Index, currently states
> Tcl 8.5 is at End of Life -- please upgrade to 9.0 (or 8.6)!!!

So they would also like users to upgrade.

Extended support for WinCE ended a year ago. I would suggest upgrading.

**Reply by RasPython3** [reply]
Thank you for your quick reply.

> Tcl 8.5 was released in 2007. So everything has had 15 years to upgrade from Tcl 8.4 before Pillow removed support for it.

Hmm, that's probably true... I had not checked how long ago when 8.5 released.

> Extended support for WinCE ended a year ago. I would suggest upgrading.

I appreciate your suggestion.
I'm Just soliloquizing; WinCE is out of support, but playing with and developping on it makes me very fun!
So I'll try to port newer version of Tcl/Tk someday.
